File: src/test/java/pl/training/module4/encapsulation/EquipmentCatalogTest.java
package pl.training.module4.encapsulation;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.math.BigDecimal;
import java.util.EnumMap;
import java.util.Map;

import org.junit.jupiter.api.Test;

import pl.training.module4.model.EquipmentType;

final class EquipmentCatalogTest {
    @Test
    void legacyCatalogSharesItsMutableMapWithTheCaller() {
        EnumMap<EquipmentType, BigDecimal> source = rates();
        LegacyEquipmentCatalog catalog = new LegacyEquipmentCatalog(
                "Summer rental",
                source);

        source.put(EquipmentType.DRILL, new BigDecimal("1.00"));

        assertEquals(
                new BigDecimal("1.00"),
                catalog.dailyRates.get(EquipmentType.DRILL));
    }

    @Test
    void accessorBasedCatalogPreservesAliasesDuringControlledMigration() {
        EnumMap<EquipmentType, BigDecimal> source = rates();
        AccessorBasedEquipmentCatalog catalog =
                new AccessorBasedEquipmentCatalog("Summer rental", source);

        source.put(EquipmentType.DRILL, new BigDecimal("1.00"));
        catalog.setName(" ");

        assertSame(source, catalog.dailyRates());
        assertEquals(
                new BigDecimal("1.00"),
                catalog.dailyRates().get(EquipmentType.DRILL));
        assertEquals(" ", catalog.name());
    }

    @Test
    void encapsulatedCatalogOwnsRatesAndReturnsUnmodifiableSnapshots() {
        EnumMap<EquipmentType, BigDecimal> source = rates();
        EquipmentCatalog catalog = new EquipmentCatalog(
                "Summer rental",
                source);
        Map<EquipmentType, BigDecimal> snapshot = catalog.dailyRates();

        source.put(EquipmentType.DRILL, new BigDecimal("1.00"));
        assertEquals(
                new BigDecimal("39.99"),
                catalog.dailyRateFor(EquipmentType.DRILL));

        catalog.changeDailyRate(
                EquipmentType.DRILL,
                new BigDecimal("42.00"));

        assertEquals(
                new BigDecimal("39.99"),
                snapshot.get(EquipmentType.DRILL));
        assertEquals(
                new BigDecimal("42.00"),
                catalog.dailyRateFor(EquipmentType.DRILL));
        assertThrows(
                UnsupportedOperationException.class,
                () -> snapshot.put(
                        EquipmentType.GENERATOR,
                        new BigDecimal("120.00")));
    }

    @Test
    void changesNameOnlyThroughValidatedOperation() {
        EquipmentCatalog catalog = new EquipmentCatalog(
                "Summer rental",
                rates());

        catalog.renameTo("Winter rental");

        assertEquals("Winter rental", catalog.name());
        assertThrows(
                IllegalArgumentException.class,
                () -> catalog.renameTo(" "));
    }

    @Test
    void normalizesRateBeforeCheckingItsInvariant() {
        EquipmentCatalog catalog = new EquipmentCatalog(
                "Summer rental",
                rates());

        catalog.changeDailyRate(
                EquipmentType.DRILL,
                new BigDecimal("42.005"));

        assertEquals(
                new BigDecimal("42.01"),
                catalog.dailyRateFor(EquipmentType.DRILL));
        assertThrows(
                IllegalArgumentException.class,
                () -> catalog.changeDailyRate(
                        EquipmentType.DRILL,
                        new BigDecimal("0.004")));
    }

    private static EnumMap<EquipmentType, BigDecimal> rates() {
        EnumMap<EquipmentType, BigDecimal> rates =
                new EnumMap<>(EquipmentType.class);
        rates.put(EquipmentType.DRILL, new BigDecimal("39.99"));
        return rates;
    }
}

File: csharp/tests/Training.Module4.Tests/Encapsulation/EquipmentCatalogTest.cs
using Training.Module4.Encapsulation;
using Training.Module4.Model;

namespace Training.Module4.Tests.Encapsulation;

public sealed class EquipmentCatalogTest
{
    [Fact]
    public void LegacyCatalogSharesItsMutableMapWithTheCaller()
    {
        var source = Rates();
        var catalog = new LegacyEquipmentCatalog("Summer rental", source);

        source[EquipmentType.Drill] = 1.00m;

        Assert.Equal(1.00m, catalog.DailyRates[EquipmentType.Drill]);
    }

    [Fact]
    public void AccessorBasedCatalogPreservesAliasesDuringControlledMigration()
    {
        var source = Rates();
        var catalog = new AccessorBasedEquipmentCatalog("Summer rental", source);

        source[EquipmentType.Drill] = 1.00m;
        catalog.Name = " ";

        Assert.Same(source, catalog.DailyRates);
        Assert.Equal(1.00m, catalog.DailyRates[EquipmentType.Drill]);
        Assert.Equal(" ", catalog.Name);
    }

    [Fact]
    public void EncapsulatedCatalogOwnsRatesAndReturnsUnmodifiableSnapshots()
    {
        var source = Rates();
        var catalog = new EquipmentCatalog("Summer rental", source);
        var snapshot = catalog.DailyRates();

        source[EquipmentType.Drill] = 1.00m;
        Assert.Equal(39.99m, catalog.DailyRateFor(EquipmentType.Drill));

        catalog.ChangeDailyRate(EquipmentType.Drill, 42.00m);

        Assert.Equal(39.99m, snapshot[EquipmentType.Drill]);
        Assert.Equal(42.00m, catalog.DailyRateFor(EquipmentType.Drill));
        var mutableView = Assert.IsAssignableFrom<IDictionary<EquipmentType, decimal>>(snapshot);
        Assert.Throws<NotSupportedException>(
            () => mutableView.Add(EquipmentType.Generator, 120.00m));
    }

    [Fact]
    public void ChangesNameOnlyThroughValidatedOperation()
    {
        var catalog = new EquipmentCatalog("Summer rental", Rates());

        catalog.RenameTo("Winter rental");

        Assert.Equal("Winter rental", catalog.Name);
        Assert.Throws<ArgumentException>(() => catalog.RenameTo(" "));
    }

    [Fact]
    public void NormalizesRateBeforeCheckingItsInvariant()
    {
        var catalog = new EquipmentCatalog("Summer rental", Rates());

        catalog.ChangeDailyRate(EquipmentType.Drill, 42.005m);

        Assert.Equal(42.01m, catalog.DailyRateFor(EquipmentType.Drill));
        Assert.Throws<ArgumentException>(
            () => catalog.ChangeDailyRate(EquipmentType.Drill, 0.004m));
    }

    private static Dictionary<EquipmentType, decimal> Rates()
    {
        return new Dictionary<EquipmentType, decimal>
        {
            [EquipmentType.Drill] = 39.99m
        };
    }
}
